### DISCUSSION TITLE
How to catch custom exception on Action button of a form ?

### DISCUSSION THREAD

**Discussion by sergeynilov** [filamentphp/filament#9334] [Help]
In laravel 10 app with filament 3.0-stable in Item editor app/Filament/Resources/ItemResource/Pages/EditItem.php I need 
to add action to make some operarion on submit on the
opened Item model, but only in case when field is_passed of Item model is true. I do:

```
<?php

namespace App\Filament\Resources\ItemResource\Pages;

use App\Exceptions\ItemMustBePassed;
use App\Filament\Resources\ItemResource;
use Carbon\Carbon;
use Filament\Forms\Components\Checkbox;
use Filament\Forms\Components\DateTimePicker;
use Filament\Forms\Components\Select;
use Filament\Forms\Components\TextInput;
...
class EditItem extends EditRecord
{
    protected static string $resource = ItemResource::class;

    protected function getHeaderActions(): array
    {
        throw_if(! $this->record->is_passed,
            new ItemMustBePassed('This item must be passed to create Item Operation !'));

        return [
            Actions\DeleteAction::make(),
            Action::make('Create Item Operation')
                ->modalIcon('heroicon-o-trash')
                ->steps([
                    Step::make('Name')
                       ...
                ])
                ->action(function (array $data, array $arguments): void {
                    ...
                })
        ];
    }

```
But the problems are :
a) when field is_passed equals false I got custom exception opening this model editor, not when clicking  on Action button
b) If there is a way to show some popup modal dialog with eror message( but using custom exception ), but not exception page on the page ?

**Comment by leandrocfe** [comment]
You should use the logic inside the actions().

1) Would you like to use `record` or a [custom form](https://filamentphp.com/docs/3.x/actions/modals/#modal-forms) ?

2) you can use [notifications](https://filamentphp.com/docs/3.x/notifications/sending-notifications/#overview)

**Reply by sergeynilov** [reply]
a) I use $this->record      when I need to use current data of the page :

`    $this->record->is_passed
`
b) I try to use Notification::make( in getHeaderActions function :
```

protected function getHeaderActions(): array
{

        if(! $this->record->is_passed) {
            Notification::make()
                ->title('This user quiz request must be passed to create meeting !')
                ->error()
                ->send();
            return [];
        }

```
But in this case I got message error when page is opened, but not when Action button is clicked as I need.
Also I do not any Action button, I suppose as I exit with
`    return [];`

If to remove line
`    return [];`

I got message twice, on page opened and button clicked, but anyway I see form on action button, which I need to hide.
In which way can I do it ?

**Reply by leandrocfe** [reply]
Not sure if this code is 100% correct because I'm writing on mobile, but let me know if you want to do something like this

```php
protected function getHeaderActions(): array
    {
        return [
            Action::make('Create Item Operation')
            ->before(function (Action $action){
 if(! $this->record->is_passed) {
            Notification::make()
                ->title('This user quiz request must be passed to create meeting !')
                ->error()
                ->send();
                
                $action->halt();
        }
            })
                ->action(function (): void {
                    
                   //create meeting...
                })
        ];
    }
```

**Reply by sergeynilov** [reply]
Looks like  :
```
           Action::make('Create Item Operation')
            ->before(function (Action $action){

```
is Triggered when use clicks on the Submit button of the modal action. I need event BEFORE modal is opened.
Also I do not see any events description at docs https://filamentphp.com/docs/3.x/actions/modals
Are there some other docs where events of Actions buttons/Modal Form are described and I need to search there ?

**Reply by leandrocfe** [reply]
Using the mount method?
```php
public function mount(): void

{

parent::mount();
//...

}
```

**Reply by sergeynilov** [reply]
Not sure how parent::mount() can be used in my case... Please give a link at docs where it is described...

**Reply by leandrocfe** [reply]
Pages are livewire components

https://filamentphp.com/docs/3.x/panels/pages/#creating-a-page

https://livewire.laravel.com/docs/lifecycle-hooks#mount
